packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -72,88 +81,145 @@ export function endTrackFrame(): Tag {
   return current!.combine();
 }
 
+export function isTracking() {
+  return CURRENT_TRACKER !== null;
+}
+
+export function consumeTag(tag: Tag) {
+  if (CURRENT_TRACKER !== null) {
+    CURRENT_TRACKER.add(tag);
+  }
+}
+
 //////////
 
-const IS_CONST_MAP: WeakMap<object, boolean | undefined> = new WeakMap();
-
-export function memo<T>(cb: () => T, context?: string | false): () => T;
-export function memo<T, U>(cb: (a1: U) => T, context?: string | false): (a1: U) => T;
-export function memo<T, U, V, W>(
-  cb: (a1: U, a2: V, a3: W) => T,
-  context?: string | false
-): (a1: U, a2: V, a3: W) => T;
-export function memo<T, U, V, W, X>(
-  cb: (a1: U, a2: V, a3: W, a4: X) => T,
-  context?: string | false
-): (a1: U, a2: V, a3: W, a4: X) => T;
-export function memo<T, U, V, W, X, Y>(
-  cb: (a1: U, a2: V, a3: W, a4: X, a5: Y) => T,
-  context?: string | false
-): (a1: U, a2: V, a3: W, a4: X, a5: Y) => T;
-export function memo<T, U, V, W, X, Y, Z>(
-  cb: (a1: U, a2: V, a3: W, a4: X, a5: Y, a6: Z) => T,
-  context?: string | false
-): (a1: U, a2: V, a3: W, a4: X, a5: Y, a6: Z) => T;
-
-export function memo<T>(callback: (...args: unknown[]) => T, debuggingContext?: string | false) {
-  let lastValue: T | undefined;
-  let tag: Tag;
-  let snapshot: Revision;
-
-  let memoized = (...args: unknown[]): T => {
-    if (!tag || !validateTag(tag, snapshot)) {
-      beginTrackFrame();
-
-      try {
-        if (DEBUG) {
-          runInAutotrackingTransaction!(() => (lastValue = callback(...args)), debuggingContext);
-        } else {
-          lastValue = callback(...args);
-        }
-      } finally {
-        tag = endTrackFrame();
-        snapshot = valueForTag(tag);
-        consumeTag(tag);
-
-        // If the final tag is constant, then we know for sure that this
-        // memoized function can never change. There are times when this
-        // information is useful externally (i.e. in the append VM, it tells us
-        // whether or not to emit opcodes) so we expose it via a metadata weakmap.
-        if (tag === CONSTANT_TAG) {
-          IS_CONST_MAP.set(memoized, true);
-        } else if (DEBUG) {
-          // In DEBUG, set the value to false explicitly. This way we can throw
-          // if someone attempts to call `isConst(memoized)` before running
-          // `memoized()` at least once.
-          IS_CONST_MAP.set(memoized, false);
-        }
-      }
-    } else {
-      consumeTag(tag);
-    }
+const CACHE_KEY: unique symbol = symbol('CACHE_KEY');
+
+export function memo<T>(callback: () => T, debuggingContext?: string | false) {
+  let cache = createCache(callback, debuggingContext);
+
+  let memoized = () => getValue(cache);
+
+  (memoized as any)[CACHE_KEY] = cache;
+
+  return memoized;
+}
+
+export function isConstMemo(memoized: Function) {
+  let cache = (memoized as any)[CACHE_KEY] as Cache | undefined;
+
+  return cache === undefined ? false : isConst(cache);
+}
```

Comment:
p1/5 Can we do the same trick that we ended up with below to ensure that TS errors if you pass a function that was not the result of calling `memo` into `isConstMemo`?